File: Android/app/src/androidTest/java/jp/co/riso/smartdeviceapp/model/printsettings/XmlNodeTest.java
package jp.co.riso.smartdeviceapp.model.printsettings;

import android.test.AndroidTestCase;
import android.util.Log;

import org.w3c.dom.Document;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.List;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import jp.co.riso.android.util.AppUtils;
import jp.co.riso.smartdeviceapp.AppConstants;
import jp.co.riso.smartdeviceapp.SmartDeviceApp;

public class XmlNodeTest extends AndroidTestCase {
    private static final String TAG = "XmlNodeTest";
    private List<XmlNode> mGroupList = new ArrayList<XmlNode>();
    private List<XmlNode> mOptionList = new ArrayList<XmlNode>();
    private List<XmlNode> mSettingList = new ArrayList<XmlNode>();

    @Override
    protected void setUp() throws Exception {
        super.setUp();
        String xmlString = AppUtils.getFileContentsFromAssets(SmartDeviceApp.getAppContext(), "printsettings.xml");

        Document printSettingsContent = null;

        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
        try {
            DocumentBuilder db = dbf.newDocumentBuilder();

            InputSource is = new InputSource();
            is.setCharacterStream(new StringReader(xmlString));
            printSettingsContent = db.parse(is);
        } catch (ParserConfigurationException | IOException | SAXException e) {
            Log.e(TAG, "Error: " + e.getMessage());
        }

        if (printSettingsContent == null) {
            return;
        }

        NodeList groupList = printSettingsContent.getElementsByTagName(XmlNode.NODE_GROUP);

        // looping through all item nodes <item>
        for (int i = 0; i < groupList.getLength(); i++) {
            XmlNode group = new Group(groupList.item(i));
            mGroupList.add(group);
        }


        NodeList optionList = printSettingsContent.getElementsByTagName("option");

        // looping through all item nodes <item>
        for (int i = 0; i < groupList.getLength(); i++) {
            XmlNode option = new Option(optionList.item(i));
            mOptionList.add(option);
        }


        NodeList settingList = printSettingsContent.getElementsByTagName("setting");

        // looping through all item nodes <item>
        for (int i = 0; i < groupList.getLength(); i++) {
            XmlNode setting = new Setting(settingList.item(i));
            mSettingList.add(setting);
        }
    }

    @Override
    protected void tearDown() throws Exception {
        super.tearDown();
    }


    public void testPreConditions() {
        assertEquals(3*AppConstants.PRINTER_TYPES.length, mGroupList.size());
    }

    public void testGetAttributeValue() {
        for (int i = 0; i < mGroupList.size(); i++) {
            XmlNode g = mGroupList.get(i);
            String attribName = g.getAttributeValue("name");
            String attribText = g.getAttributeValue("text");
            switch(i%(AppConstants.PRINTER_TYPES.length-2)) {
                case 0:
                    assertEquals("basic", attribName);
                    assertEquals("ids_lbl_basic", attribText);
                    break;
                case 1:
                    assertEquals("layout", attribName);
                    assertEquals("ids_lbl_layout", attribText);
                    break;
                case 2:
                    assertEquals("finishing", attribName);
                    assertEquals("ids_lbl_finishing", attribText);
                    break;
                default:
                    fail("invalid group");
            }
        }

        assertEquals("", mOptionList.get(0).getAttributeValue("name"));
        assertEquals("colorMode", mSettingList.get(0).getAttributeValue("name"));
    }
    
    public void testGetAttributeValue_NullValues() {
        assertEquals("", mGroupList.get(0).getAttributeValue(null));
        assertEquals("", mOptionList.get(0).getAttributeValue(null));
        assertEquals("", mSettingList.get(0).getAttributeValue(null));
    }

}

File: Android/app/src/androidTest/java/jp/co/riso/smartdeviceapp/model/printsettings/XmlNodeTest.kt
package jp.co.riso.smartdeviceapp.model.printsettings

import android.util.Log
import jp.co.riso.android.util.AppUtils
import jp.co.riso.smartdeviceapp.AppConstants
import jp.co.riso.smartdeviceapp.SmartDeviceApp
import junit.framework.TestCase
import org.junit.Before
import org.junit.Test
import org.w3c.dom.Document
import org.xml.sax.InputSource
import org.xml.sax.SAXException
import java.io.IOException
import java.io.StringReader
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException

class XmlNodeTest {

    private val _groupList: MutableList<XmlNode> = ArrayList()
    private val _optionList: MutableList<XmlNode> = ArrayList()
    private val _settingList: MutableList<XmlNode> = ArrayList()

    @Before
    fun setUp() {
        val xmlString =
            AppUtils.getFileContentsFromAssets(SmartDeviceApp.getAppContext(), "printsettings.xml")
        var printSettingsContent: Document? = null
        val dbf = DocumentBuilderFactory.newInstance()
        try {
            val db = dbf.newDocumentBuilder()
            val `is` = InputSource()
            `is`.characterStream = StringReader(xmlString)
            printSettingsContent = db.parse(`is`)
        } catch (e: ParserConfigurationException) {
            Log.e(TAG, "Error: " + e.message)
        } catch (e: IOException) {
            Log.e(TAG, "Error: " + e.message)
        } catch (e: SAXException) {
            Log.e(TAG, "Error: " + e.message)
        }
        if (printSettingsContent == null) {
            return
        }
        val groupList = printSettingsContent.getElementsByTagName(XmlNode.NODE_GROUP)

        // looping through all item nodes <item>
        for (i in 0 until groupList.length) {
            val group: XmlNode = Group(groupList.item(i))
            _groupList.add(group)
        }
        val optionList = printSettingsContent.getElementsByTagName("option")

        // looping through all item nodes <item>
        for (i in 0 until groupList.length) {
            val option: XmlNode = Option(optionList.item(i))
            _optionList.add(option)
        }
        val settingList = printSettingsContent.getElementsByTagName("setting")

        // looping through all item nodes <item>
        for (i in 0 until groupList.length) {
            val setting: XmlNode = Setting(settingList.item(i))
            _settingList.add(setting)
        }
    }

    @Test
    fun testPreConditions() {
        TestCase.assertEquals(3 * AppConstants.PRINTER_TYPES.size, _groupList.size)
    }

    @Test
    fun testGetAttributeValue() {
        for (i in _groupList.indices) {
            val g = _groupList[i]
            val attribName = g.getAttributeValue("name")
            val attribText = g.getAttributeValue("text")
            when (i % (AppConstants.PRINTER_TYPES.size - 2)) {
                0 -> {
                    TestCase.assertEquals("basic", attribName)
                    TestCase.assertEquals("ids_lbl_basic", attribText)
                }
                1 -> {
                    TestCase.assertEquals("layout", attribName)
                    TestCase.assertEquals("ids_lbl_layout", attribText)
                }
                2 -> {
                    TestCase.assertEquals("finishing", attribName)
                    TestCase.assertEquals("ids_lbl_finishing", attribText)
                }
                else -> TestCase.fail("invalid group")
            }
        }
        TestCase.assertEquals("", _optionList[0].getAttributeValue("name"))
        TestCase.assertEquals("colorMode", _settingList[0].getAttributeValue("name"))
    }

    @Test
    fun testGetAttributeValue_NullValues() {
        TestCase.assertEquals("", _groupList[0].getAttributeValue(null))
        TestCase.assertEquals("", _optionList[0].getAttributeValue(null))
        TestCase.assertEquals("", _settingList[0].getAttributeValue(null))
    }

    companion object {
        private const val TAG = "XmlNodeTest"
    }
}